src/top-bar/org-unit-select/approval-statuses.js

```diff
@@ -0,0 +1,165 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import React, { createContext, useContext, useRef, useState } from 'react'
+import { useDebouncedCallback } from 'use-debounce'
+
+const ApprovalStatusesContext = createContext()
+
+const useApprovalStatusesContext = () => useContext(ApprovalStatusesContext)
+
+class ApprovalStatusesMap {
+    constructor(map) {
+        this._map = map || new Map()
+    }
+
+    _serialiseKey({ workflowId, periodId, orgUnitId }) {
+        return `${workflowId}-${periodId}-${orgUnitId}`
+    }
+
+    set(key, status) {
+        this._map.set(this._serialiseKey(key), status)
+    }
+
+    get(key) {
+        return this._map.get(this._serialiseKey(key))
+    }
+
+    clone() {
+        return new ApprovalStatusesMap(new Map(this._map))
+    }
+}
+
+export const ApprovalStatusesProvider = ({ children }) => {
+    const [approvalStatuses, setApprovalStatuses] = useState(
+        new ApprovalStatusesMap()
+    )
+
+    const updateApprovalStatuses = ({
+        workflowId,
+        periodId,
+        approvalStatusUpdates,
+    }) => {
+        setApprovalStatuses(approvalStatuses => {
+            const newApprovalStatuses = approvalStatuses.clone()
+            for (const [orgUnitId, status] of Object.entries(
+                approvalStatusUpdates
+            )) {
+                newApprovalStatuses.set(
+                    { workflowId, periodId, orgUnitId },
+                    status
+                )
+            }
+            return newApprovalStatuses
+        })
+    }
+
+    return (
+        <ApprovalStatusesContext.Provider
+            value={{ approvalStatuses, updateApprovalStatuses }}
+        >
+            {children}
+        </ApprovalStatusesContext.Provider>
+    )
+}
+
+// Debounced version
+const useFetchApprovalStatuses = ({ updateApprovalStatuses }) => {
```

Comment:
The aim of this debounced version of `useFetchApprovalStatuses` is to reduce the number of requests to the `dataApprovals/approvals` endpoint by batching them up.

As `fetchApprovalStatuses` is called in the `onChildrenLoaded` callback of the org unit tree, this debouncing is not very useful in higher-latency connections. Implementing and using a `onWillLoadChildren` prop on the org unit tree would allow us to leverage this debouncing better.

Reply:
I think what you've done here makes sense in principle: it seems a shame to fire a bunch of requests almost at the same time when those requests could all be combined into one single request.

But I am not completely sure that debounce is the way to tackle this issue. In fact and not even sure that we are currently leveraging the most appropriate callback prop. By using `onChildrenLoaded` we are effectively pre-fetching all child statuses for nodes that aren't even visible yet. Some might never become visible at all, if the user doesn't open the parent node. So even though we cache responses and we batch requests we are still wasting a lot of resources this way

In theory, the `onExpand` prop seems a better fit, because it would enable us to load theses statuses "on-demand". However, there currently are two problems with this:
1. We don't get access to the full node, but just the `path` in the callback
2. This callback only fires on user-triggered expands, i.e. when a click happens to open a node. It doesn't get called when nodes are expanding as a result of the presence of an `initiallyExpanded` prop.

I would probably consider point 1 to be an additional feature in the org-unit-tree in UI). Point two I would probably classify as a bug, but changing it could possibly break existing apps. So I guess we'd better discuss this with the team first.